### PR TITLE
Fixing ipam concurrent allocation

### PR DESCRIPTION
**Pull request by abhinandanpb**
Adding locks to ensure concurrent bitseq allocation doesnt corrupt the RLE encode bitseq.
It was seen with a small test that concurrent use of ipam address request was allocating duplicate IP
addresses

Signed-off-by: Abhinandan Prativadi <abhi@docker.com>

### CODE REVIEW

File: bitseq/sequence.go [hunk 1]
```diff
@@ -339,40 +339,30 @@ func (h *Handle) set(ordinal, start, end uint64, any bool, release bool) (uint64
 		}
 
 		// Create a private copy of h and work on it
```
Comment (fcrisciani) [inline bitseq/sequence.go]: comment not anymore valid

File: bitseq/sequence.go [hunk 2]
```diff
@@ -308,17 +308,17 @@ func (h *Handle) set(ordinal, start, end uint64, any bool, release bool) (uint64
...
+		h.RUnlock()
 		if store != nil {
-			if err := store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
+			if err = store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
 				return ret, err
 			}
 		}
 
-		h.Lock()
 		// Get position if available
```
Comment (fcrisciani) [inline bitseq/sequence.go]: I see the following problem here.
Go routine A takes from the store objA
Go routine B takes from the store objA (using the same key)

Go routine A now modifies objA, same the Go routine B with its own copy

Now Go routine A saves its object into the store
Go routine B saves its object into the store and rewrites the obj of Go routine A

At this point the state of Go routine A is lost, I think we need a way to guarantee that 2 go routines won't fetch the same object from the store

Comment (abhinandanpb) [inline bitseq/sequence.go]: Yep makes sense.

**Comment by fcrisciani** [comment]
It can work only if the store returns a pointer to the very same object so
that the 2 go routines are not using an outdated one

File: bitseq/sequence.go (hunk 2 above)
Comment (mavenugo) [inline bitseq/sequence.go]: @fcrisciani thats correct. But in this case, the `set` method already handles this scenario by retrying the entire operation if the AtomicPut fails in the store. Also the Store's AtomicPut guarantees the write consistency with the use of Object Index. If an object that is writing to the store has an older data Index, then the write will fail and the `set` will retry the operation. Please look the logic below in this method, where we handle the `types.RetryError` and redo the entire `set` again.

**Comment by mavenugo** [comment]
@fcrisciani as explained above, I dont think that is an issue with this particular function. `set` takes the help of AtomicWrite guarantees of the store.

**Comment by fcrisciani** [comment]
ok then LGTM

File: bitseq/sequence.go [hunk 3]
```diff
@@ -262,19 +264,17 @@ func (h *Handle) runConsistencyCheck() bool {
...
-		h.Unlock()
+		h.RUnlock()
 
 		if store != nil {
 			if err := store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
 				return err
 			}
 		}
 
-		h.Lock()
```
Comment (mavenugo) [inline bitseq/sequence.go]: Isnt it correct to get the Lock for `getCopy()` ?

File: bitseq/sequence.go [hunk 4]
```diff
@@ -308,17 +308,17 @@ func (h *Handle) set(ordinal, start, end uint64, any bool, release bool) (uint64
...
 		if store != nil {
-			if err := store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
+			if err = store.GetObject(datastore.Key(h.Key()...), h); err != nil && err != datastore.ErrKeyNotFound {
 				return ret, err
 			}
 		}
 
-		h.Lock()
 		// Get position if available
+		h.Lock()
```
Comment (mavenugo) [inline bitseq/sequence.go]: eyesore. pls move it above the comment to avoid unnecessary changes in the diff.

File: bitseq/sequence.go (hunk 3 above)
Comment (abhinandanpb) [inline bitseq/sequence.go]: getCopy will acquire the required lock.

Comment (mavenugo) [inline bitseq/sequence.go]: 👍

**Comment by fcrisciani** [comment]
it should because 2 go routines will get the same copy and will set the very same bit, the store I don't think will have any way to distinguish it so will allow it, but instead 2 go routines have assigned the same bit. Makes sense?

**Comment by abhinandanpb** [comment]
@mavenugo I agree on that front. Closing
